atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -678,6 +732,21 @@ public ListenableFuture<Map<Cell, Value>> getAsync(TableReference tableRef, Map<
         return Futures.immediateFuture(get(tableRef, timestampByCell));
     }
 
+    /**
+     * This takes out a reentrant read lock. It is only exclusive specifically with
+     * {@link #multiCheckAndSet(MultiCheckAndSetRequest)}.
+     */
+    private <T> T runWithLockForWrites(TableReference tableRef, Function<ConcurrentSkipListMap<Key, byte[]>, T> task) {
```

Comment:
Hmm the name is confusing 👀  Also why do we not need the read lock for read operations?